docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,293 @@
+﻿# Toplevel inference
+
+Owner: leafp@google.com
+
+Status: Proposal
+
+
+## Top level inference overview
+
+Top level inference has two phases:
+1. Method inference
+2. Static variable and field inference
+
+
+All results from the first phase are available in the second phase.
+
+
+As a general principle, when inference fails it is an error.  Tools are free to
+behave in implementation specific ways to provide a graceful user experience,
+but with respect to the language and its semantics, programs for which inference
+fails are unspecified.
+
+
+The intuitive idea behind this proposal is that top level inference works by
+first inferring methods, and then doing inference for "field like things"
+(static variables, fields, setters, and getters), where we don't allow inference
+for "field like things" to depend on the results of inference for "field like
+things" except in cases where the dependencies are statically visible.
+
+
+Some broad principles for type inference that the language team has agreed to,
+and which this proposal is designed to satisfy:
+* Type inference should be free to fail with an error, rather than being
+  required to always produce some answer
+* It should not be possible for a programmer to observe a declaration as having
+  two different types (e.g. dynamic for recursive uses, but subsequently int).
+  Some consistent answer (or an error) should always be produced.
+* The inference for local variables, toplevel variables, and fields, should
+  either agree or error out.  The same expression should not be inferred
+  differently at different syntactic positions.  It’s ok for an expression to be
+  inferrable at one level but not at another.
+* Obvious types should be inferred
+* Inferred and annotated types should be treated the same
```

Comment:
I like the principle because when I look at examples where we don't do this, we seem to get stranger behavior (see my example on your previous comment for example).

I would prefer to resolve this by doing less signature inference, but I'm still somewhat persuaded that insofar as we are going to do signature inference, the least surprising thing is to do it as uniformly as possible.